UDP firewall flood checking

neoaggelos opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #2079

#### Changes
<!-- What are the changes made in this pull request? -->

- Add a circular buffer of timestamps on UDP firewall per gateway
- Drop messages that exceed a specified rate limit.

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

- Naming is all over the place, please suggest better names where possible
- Current limits are hard-coded in the code. Should probably be GS configuration options?
- Relevant tests have been added.
- Circular buffer implementation is based on `./pkg/gatewayserver/rtts.go`. In this case, maybe mutexes are not required at all?

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [X] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

johanstokking commented on pkg/gatewayserver/io/udp/firewall.go:
```diff
@@ -25,25 +25,50 @@ import (
+	// Configure age threshold for oldest message when rate limiting. If oldest
+	// message is not older than this duration, then the rate is considered to
+	// have been exceeded and the packet is dropped.
+	defaultRateLimitingAgeThreshold = 10 * time.Millisecond
```
Comments start with the symbol name

johanstokking replied:
And a `const` is not configuration

neoaggelos commented:
Changes since review:

- Refactored into a `rateLimitingFirewall`, which can wrap an existing `Firewall` while also providing the rate limiting capabilities.
- Rate limiting now takes precedence over the wrapped firewall, which is much better (during the flood, we avoid doing more stuff than needed).
- Added configuration options on the GS (`--gs.udp.rate-limiting`).
- Updated circular buffer implementation to use a fixed size array and a running pointer.
- Updated documentation and changelog

Will squash commits after reviewing is done.

neoaggelos commented on pkg/gatewayserver/io/udp/firewall_ratelimit.go:
```diff
@@ -0,0 +1,94 @@
+
+	// defaultThreshold is the duration after which a new packet will be accepted.
+	defaultThreshold = 10 * time.Millisecond
+)
```
These are probably not needed

johanstokking replied:
Indeed, they can be removed. GS has default config already

neoaggelos commented on pkg/gatewayserver/io/udp/firewall_ratelimit.go:
```diff
@@ -0,0 +1,93 @@
+	if ok {
+		timestamps = val.(*incoming)
+		oldestTimestamp := timestamps.Oldest()
+		if !oldestTimestamp.IsZero() && now.Sub(oldestTimestamp) < f.threshold {
```
The `IsZero()` check is probably not needed, since if that is true, then the second condition will most likely fail

neoaggelos commented on pkg/gatewayserver/io/udp/udp.go:
```diff
@@ -61,6 +71,11 @@ var DefaultConfig = Config{
 	ScheduleLateTime:    800 * time.Millisecond,
 	AddrChangeBlock:     1 * time.Minute, // Release address when the connection expires.
+	RateLimiting: RateLimitingConfig{
+		Enable:    true,
```
We might not want to enable this by default

johanstokking replied:
We actually do

neoaggelos commented on pkg/gatewayserver/io/udp/udp.go:
```diff
@@ -32,6 +32,14 @@ import (
+type RateLimitingConfig struct {
+	Enable    bool          `name:"enable" description:"Enable rate limiting for gateways"`
+	Messages  int           `name:"messages" description:"Number of past messages to check timestamp for"`
+	Threshold time.Duration `name:"threshold" description:"Filter packet if timestamp is not newer than the older timestamps of the previous messages by this threshold"`
```
Not too good at naming new things, these may need to change

johanstokking replied:
Name is good

johanstokking commented on doc/content/reference/configuration/gateway-server.md:
```diff
@@ -62,3 +62,10 @@ Using the `packet-buffer` and `packet-handlers` options, the throughput of UDP p
+The Gateway Server supports rate limiting traffic for gateways that are using the Semtech UDP protocol.
+
+- `gs.udp.rate-limiting.enable`: Enable rate limiting for gateways.
+- `gs.udp.rate-limiting.messages`: Number of past messages to check timestamp for
```
Trailing dot

johanstokking commented on pkg/gatewayserver/io/udp/firewall_ratelimit.go:
```diff
@@ -0,0 +1,93 @@
+type rateLimitingFirewall struct {
+	f Firewall
+
+	m sync.Map // eui to incoming
```
`eui` is not a type

neoaggelos replied:
Updated comment

johanstokking commented on pkg/gatewayserver/io/udp/firewall_ratelimit.go:
```diff
@@ -0,0 +1,93 @@
+	}
+	if threshold == 0 {
+		threshold = defaultThreshold
+	}
```
These checks are useful in libraries with a public API

If there's sanity checking for bad config, it should return an error here. But I don't think that is all necessary and that you can trust the values

neoaggelos replied:
Removed, GS default config should suffice

johanstokking commented on pkg/gatewayserver/io/udp/firewall_ratelimit.go:
```diff
@@ -0,0 +1,93 @@
+
+	// Update firewall
+	timestamps.Append(now)
+	f.m.Store(eui, timestamps)
```
There's no need to update the map. You probably just want to store a pointer to the timestamps

neoaggelos replied:
Updated to only store the reference once (when creating the new object)

johanstokking commented on pkg/gatewayserver/io/udp/firewall_test.go:
```diff
@@ -148,4 +149,57 @@ func TestMemoryFirewall(t *testing.T) {
+
+		for _, tc := range []struct {
+			rateLimiting bool
+			verify       func(error) bool
```
`errorAssertion`

johanstokking commented on pkg/gatewayserver/io/udp/firewall_test.go:
```diff
@@ -148,4 +149,57 @@ func TestMemoryFirewall(t *testing.T) {
+		}{
+			{
+				rateLimiting: true,
+				verify:       func(err error) bool { return errors.IsResourceExhausted(err) },
```
`errorAssertion: errors.IsResourceExhausted`

johanstokking commented on pkg/gatewayserver/io/udp/firewall_test.go:
```diff
@@ -148,4 +149,57 @@ func TestMemoryFirewall(t *testing.T) {
+				verify:       func(err error) bool { return err == nil },
+			},
+		} {
+			t.Run(fmt.Sprintf("RateLimiting=%v", tc.rateLimiting), func(t *testing.T) {
```
High coupling; I would prefer keeping the tests for rate limiting separate

neoaggelos replied:
Will move into separate `firewall_ratelimiting_test.go`

johanstokking commented on pkg/gatewayserver/io/udp/firewall_ratelimit.go:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	// Update firewall
+	timestamps.Append(now)
```
`Oldest()` is a read, `Append()` a write. Both are independently locked, but it's nicer to have one atomic operation that appends and returns the delta with the oldest value.

neoaggelos replied:
Updated `Append()` to also return oldest value.

johanstokking commented on pkg/gatewayserver/io/udp/incoming.go:
```diff
@@ -0,0 +1,49 @@
+	"time"
+)
+
+type incoming struct {
```
Isn't this just `timestamps`?

neoaggelos replied:
Renamed to timestamps

johanstokking commented on pkg/gatewayserver/io/udp/incoming.go:
```diff
@@ -0,0 +1,49 @@
+)
+
+type incoming struct {
+	count   int
```
What is the purpose of this var?

neoaggelos replied:
Used to reset counter. We can drop this if `len(items)` is O(1).

johanstokking replied:
It is

neoaggelos commented:
We still need to make sure that stale entries on this table are removed at regular intervals. One way would be to have another `gc()` here like with the `memoryFirewall` and empty the map every X minutes.

A slightly better approach could be to extend the `Firewall` interface with a `Clean(ttnpb.GatewayIdentifiers)` function. This is out of scope for the existing PR.

@johanstokking thoughts?

johanstokking commented on pkg/gatewayserver/io/udp/timestamps.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+type timestamps struct {
+	count   int
```
Remove, see https://blog.golang.org/go-slices-usage-and-internals

johanstokking commented on pkg/gatewayserver/io/udp/timestamps.go:
```diff
@@ -0,0 +1,45 @@
+
+type timestamps struct {
+	count   int
+	current int
```
This is not used as `current`, it's used as the (new) position, i.e. `pos` or `position`

johanstokking commented on pkg/gatewayserver/io/udp/timestamps.go:
```diff
@@ -0,0 +1,45 @@
+func (i *timestamps) Append(val time.Time) time.Time {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	result := i.items[i.current]
```
See here; the oldest is "current", that can't be right
